Design/Crystallography/HKL.md

```diff
@@ -0,0 +1,121 @@
+# Making HKL different
+
+## Motivation
+
+Currently the class `Kernel::V3D` is used to represent a large range of points and vectors in different spaces. Detector coordinates, unit cell vectors, Miller indices and many more are all represented by the same very low-level class `V3D`. While this is certainly very flexible in terms of the number of classes and/or interfaces developers need to keep track of, it comes with a number of disadvantages. The most important is probably that it's possible to perform nonsensical operations such as adding detector coordinates to a unit cell vector - since different types do not exist, they can not be enforced by the compiler, which means the burden of preventing errors falls to the developer. A similar case is where real and reciprocal space vectors have to be transformed by using different matrices. Without dedicated types the developer needs to make sure the correct matrix is used and in addition runtime checks have to be performed.
+
+While it is a big task to make each and every use of `V3D` (and connected to that also `Kernel::Matrix`) type-safe, implementing a dedicated type for Miller indices HKL is a rather confined task. Potential replacements for `V3D` in this domain can exist in parallel at first and code can be ported step-by-step, simply reverting to the current "unsafe" implementation where necessary.
+
+## Requirements
+
+There should be a datatype different from `V3D` but with similar capabilities that is used to represent Miller index triplets.
+
+While in theory, Miller indices are defined to be composed of three integer numbers, some flexibility is required for cases in early steps of data processing where exact integers may not be known or for commensurate superstructures where fractional indices can occur.
+
+Though similar, those three cases cover different uses and should be covered by different types:
+
+  1. `HKL` (integer, "proper Miller indices")
+  2. `ProHKL` (double, "early stages of reduction")
+  3. `FractionalHKL` (double, use for magnetic or other super-structures)
```

Comment:
You're right, we could use truly rational numbers for commensurate cases, `boost::rational` is already used for `SymmetryOperation` for example. One would need to keep one eye on performance though, because `boost::rational` does things like calculating the GCD and so on in some (or all of them? not entirely sure) of the arithmetic operations.